sv.c

```diff
@@ -4771,6 +4776,272 @@ Perl_sv_setsv_flags(pTHX_ SV *dsv, SV* ssv, const I32 flags)
+     * simplification. Perhaps this might be worth revisiting in the future.
+     * -- April 2025. */
+    assert(ssv);
+    svtype stype = SvTYPE(ssv);
+
+#if NVSIZE <= IVSIZE
+    ASSUME(stype > SVt_NV);
+#else
+    ASSUME(stype > SVt_IV);
+#endif
+    ASSUME(stype <= SVt_PVMG);
+    SV *dsv = newSV_type(stype);
+    U32 sflags   = SvFLAGS(ssv);
+
+    ASSUME(!(flags & SV_GMAGIC)); /* Caller should have sorted this. */
+    /* Note: Perl_sv_setsv_flags essentially ignores magic, except for
+     * taint and vstring magic. */
+
+    /* The switch cases are in the order that gcov of make/make test suggests. */
+    switch(sflags & (SVp_IOK|SVp_NOK|SVf_ROK|SVp_POK|SVf_FAKE|SVppv_STATIC)) {
+        case SVp_IOK:
+            SvIV_set(dsv, SvIVX(ssv));
+            SvFLAGS(dsv) |= sflags & (SVf_IOK|SVp_IOK|SVf_IVisUV);
+            break;
+        case SVp_POK:
+            if (UNLIKELY(SvTYPE(ssv) == SVt_INVLIST)) {
+               invlist_clone(ssv, dsv);
+               return dsv;
+            }
+            S_sv_freshcopy_POK(aTHX_ dsv, ssv, flags);
+        {
+            const char *vstr_pv;
+            STRLEN vstr_len;
+            if ((vstr_pv = SvVSTRING(ssv, vstr_len))) {
+                sv_magic(dsv, NULL, PERL_MAGIC_vstring, vstr_pv, vstr_len);
+                SvRMAGICAL_on(dsv);
+            }
+        }
+            break;
+        case SVp_POK|SVp_IOK:
+            S_sv_freshcopy_POK(aTHX_ dsv, ssv, flags);
+        {
+            const char *vstr_pv;
+            STRLEN vstr_len;
+            if ((vstr_pv = SvVSTRING(ssv, vstr_len))) {
+                sv_magic(dsv, NULL, PERL_MAGIC_vstring, vstr_pv, vstr_len);
+                SvRMAGICAL_on(dsv);
+            }
+        }
+            SvIV_set(dsv, SvIVX(ssv));
+            SvFLAGS(dsv) |= (sflags & (SVf_IOK|SVp_IOK|SVf_IVisUV));
+            if ((sflags & SVf_IOK) && !(sflags & SVf_POK)) {
+                /* Source was SVf_IOK|SVp_IOK|SVp_POK but not SVf_POK, meaning
+                   a value set as an integer and later stringified. So mark
+                   destination the same: */
+                SvFLAGS(dsv) &= ~SVf_POK;
+            }
+            break;
+        case SVp_POK|SVp_IOK|SVp_NOK|SVppv_STATIC: /* e.g. PL_sv_yes, PL_sv_no */
+            assert(SvTYPE(ssv) != SVt_INVLIST);
+
+            ASSUME(!SvPVX_const(dsv));
+            ASSUME(!SvOK(dsv));
+            ASSUME(!(SvFLAGS(dsv) & SVf_BREAK));
+{
+            const STRLEN cur = SvCUR(ssv);
+
+            (void)SvPOK_only(dsv);
+            SvPV_set(dsv, SvPVX(ssv));
+            SvLEN_set(dsv, 0);
+            SvCUR_set(dsv, cur);
+}
+            SvIV_set(dsv, SvIVX(ssv));
+            SvNV_set(dsv, SvNVX(ssv));
+            SvFLAGS(dsv) |= SVf_IsCOW|SVppv_STATIC|
+                (sflags & (SVf_IOK|SVp_IOK|SVf_IVisUV|SVf_NOK|SVp_NOK|SVf_POK|SVf_POK|SVf_UTF8));
+
+        {
+            const char *vstr_pv;
+            STRLEN vstr_len;
+            if ((vstr_pv = SvVSTRING(ssv, vstr_len))) {
```

Comment:
This VSTRING test is starting to look like excessive copy paste. It can't be in so many case statements.

Is this also true?
```
if (SvVSTRING(&PL_sv_undef, unused_len))
    warn("true');
```

Reply:
> This VSTRING test is starting to look like excessive copy paste. 

As per the other comment, I'll try to pull that out into the caller(s).


> Is this also true?
> 
> ```
> if (SvVSTRING(&PL_sv_undef, unused_len))
>     warn("true");
> ```

It certainly shouldn't be. `PL_sv_undef` is initialised as an `SVt_PVNV` with only these flags set: `SVf_READONLY|SVf_PROTECT|SVt_NULL`.